### PR TITLE
[ENH]: add apply_function method to epochs and evoked objects

### LINKED ISSUE #8346: FR: add apply_function method to epochs and evoked objects
based on [gitter thread](https://gitter.im/mne-tools/mne-python?at=5f7b310b17377c6b5ddcf5a3). I don't have the time to make a PR right now, but I think an issue to track this feature is a good idea.

# problem

I want to modify the data stored in epochs in some arbitrary way, so currently I use:

- get the data `X = epochs.get_data()`
- modify it `X = ...`, for example by doing a matrix multiplication with a covariance matrix of shape (nchs, nchs)
- reassign it `epochs._data = X`

but this is unreliable, because the `._data` attribute is private and not guaranteed to work this way in all cases.

# solution

Adding an [`apply_function`](https://mne.tools/dev/generated/mne.io.RawArray.html#mne.io.RawArray.apply_function) method like in `raw` objects to `epochs` and `evoked` would solve this problem.

### PR DESCRIPTION
**Pull request by vpeterson**
#### Reference issue
Fixes #8346.

#### What does this implement/fix?
first proposed version on how apply_function can be added to epochs and evoked objects. At the moment, just implementing the function to epochs objects. This follows the apply_function implemented in raw objects.

#### Additional information
CC with @nordme

### CODE REVIEW

**Comment by vpeterson** [comment]
just notice: why a commit for another branch is being shown in this PR?

**Comment by agramfort** [comment]
@vpeterson can you have a look at the CI issues? thcx

File: mne/epochs.py [hunk 1]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
         """
         return self._get_data(picks=picks, item=item)
 
+    @verbose
+    def apply_function(self, fun, picks=None, dtype=None, n_jobs=1,
+                       channel_wise=True, verbose=None, *args, **kwargs):
+        """Apply a function to a subset of channels.
+
+        The function "fun" is applied to the channels defined in "picks". The
+        data of the Epochs object is modified inplace. If the function returns
```
Comment (wmvanvliet) [inline mne/epochs.py:1502]:
```suggestion
        data of the Epochs object is modified in-place. If the function returns
```

File: mne/epochs.py [hunk 2]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
         """
         return self._get_data(picks=picks, item=item)
 
+    @verbose
+    def apply_function(self, fun, picks=None, dtype=None, n_jobs=1,
+                       channel_wise=True, verbose=None, *args, **kwargs):
+        """Apply a function to a subset of channels.
+
+        The function "fun" is applied to the channels defined in "picks". The
```
Comment (wmvanvliet) [inline mne/epochs.py:1501]:
```suggestion
        The function ``fun`` is applied to the channels defined in ``picks``. The
```

File: mne/epochs.py [hunk 3]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+        data of the Epochs object is modified inplace. If the function returns
+        a different data type (e.g. numpy.complex) it must be specified using
+        the dtype parameter, which causes the data type used for representing
+        the raw data to change.
+        The Epochs object has to have the data loaded e.g. with
+        ``preload=True`` or ``self.load_data()``.
+
+        .. note:: If n_jobs > 1, more memory is required as
```
Comment (wmvanvliet) [inline mne/epochs.py:1509]:
```suggestion
        .. note:: If ``n_jobs`` > 1, more memory is required as
```

File: mne/epochs.py [hunk 4]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+                  ``len(picks) * n_times`` additional time points need to
+                  be temporaily stored in memory.
+        .. note:: If the data type changes (dtype != None), more memory is
```
Comment (wmvanvliet) [inline mne/epochs.py:1512]:
```suggestion
        .. note:: If the data type changes (``dtype != None``), more memory is
```

File: mne/epochs.py [hunk 5]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+        Parameters
+        ----------
+        fun : callable
+            A function to be applied to the channels. The first argument of
+            fun has to be a timeseries (numpy.ndarray). The function must
+            operate on an array of shape ``(n_times,)`` if
+            ``channel_wise=True`` and ``(len(picks), n_times)`` otherwise.
+            The function must return an ndarray shaped like its input.
+        %(picks_all_data_noref)s
+        dtype : numpy.dtype (default: None)
```
Comment (wmvanvliet) [inline mne/epochs.py:1525]:
```suggestion
        dtype : numpy.dtype
```

We don't document defaults like this. Instead, we add it to the description of the parameter (see below).

File: mne/epochs.py [hunk 6]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+            Data type to use for raw data after applying the function. If None
+            the data type is not modified.
```
Comment (wmvanvliet) [inline mne/epochs.py:1527]:
```suggestion
            the data type is not modified. Defaults to ``None``.
```

File: mne/epochs.py [hunk 7]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+        n_jobs : int (default: 1)
+            Number of jobs to run in parallel. Ignored if ``channel_wise`` is
+            False.
```
Comment (wmvanvliet) [inline mne/epochs.py:1530]:
```suggestion
            False. Defaults to 1.
```

File: mne/epochs.py [hunk 8]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+            A function to be applied to the channels. The first argument of
+            fun has to be a timeseries (numpy.ndarray). The function must
+            operate on an array of shape ``(n_times,)`` if
+            ``channel_wise=True`` and ``(len(picks), n_times)`` otherwise.
+            The function must return an ndarray shaped like its input.
+        %(picks_all_data_noref)s
+        dtype : numpy.dtype (default: None)
+            Data type to use for raw data after applying the function. If None
+            the data type is not modified.
+        n_jobs : int (default: 1)
```
Comment (wmvanvliet) [inline mne/epochs.py:1528]:
```suggestion
        n_jobs : int
```

File: mne/epochs.py [hunk 9]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+            Number of jobs to run in parallel. Ignored if ``channel_wise`` is
+            False.
+        channel_wise : bool (default: True)
+            Whether to apply the function to each channel individually. If
+            False, the function will be applied to all channels at once.
```
Comment (wmvanvliet) [inline mne/epochs.py:1533]:
```suggestion
            False, the function will be applied to all channels at once.
            Defaults to ``True``.
```

File: mne/epochs.py [hunk 10]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+            ``channel_wise=True`` and ``(len(picks), n_times)`` otherwise.
+            The function must return an ndarray shaped like its input.
+        %(picks_all_data_noref)s
+        dtype : numpy.dtype (default: None)
+            Data type to use for raw data after applying the function. If None
+            the data type is not modified.
+        n_jobs : int (default: 1)
+            Number of jobs to run in parallel. Ignored if ``channel_wise`` is
+            False.
+        channel_wise : bool (default: True)
```
Comment (wmvanvliet) [inline mne/epochs.py:1531]:
```suggestion
        channel_wise : bool
```

File: mne/epochs.py [hunk 11]
```diff
@@ -1491,6 +1493,85 @@ def get_data(self, picks=None, item=None):
...
+            Whether to apply the function to each channel individually. If
+            False, the function will be applied to all channels at once.
+        %(verbose_meth)s
+        *args : list
+            Additional positional arguments to pass to fun (first pos. argument
+            of fun is the timeseries of a channel).
+        **kwargs : dict
+            Keyword arguments to pass to fun.
```
Comment (wmvanvliet) [inline mne/epochs.py:1539]:
```suggestion
            Additional keyword arguments to pass to fun.
```

**Review by wmvanvliet — CHANGES_REQUESTED** [review]
Just a few nitpicks. Looks very nice!

**Comment by vpeterson** [comment]
> Just a few nitpicks. Looks very nice!

@wmvanvliet thank you!

File: mne/epochs.py [hunk 12]
```diff
@@ -1491,6 +1493,86 @@ def get_data(self, picks=None, item=None):
         """
         return self._get_data(picks=picks, item=item)
 
+    @verbose
+    def apply_function(self, fun, picks=None, dtype=None, n_jobs=1,
+                       channel_wise=True, verbose=None, *args, **kwargs):
+        """Apply a function to a subset of channels.
+
+        The function ``fun`` is applied to the channels defined in ``picks``. The
```
Comment (larsoner) [inline mne/epochs.py:1501]:
```
Error: ./mne/epochs.py:1501:80: E501 line too long (81 > 79 characters)
```

**Comment by vpeterson** [comment]
We added the apply_function now also to evoked object with @nordme.
Let us know how the changes look like @larsoner @wmvanvliet

File: mne/evoked.py [hunk 13]
```diff
@@ -151,6 +153,83 @@ def data(self, data):
...
+
+        The function ``fun`` is applied to the channels defined in ``picks``.
+        The data of the Evoked object is modified in place. If the function
+        returns a different data type (e.g. numpy.complex) it must be specified
+        using the dtype parameter, which causes the data type used for
+        representing the raw data to change. Since evoked data is averaged over
+        time, the function defined by the user is always applied channel-wise.
+
+        The Evoked object has to have the data loaded e.g. with
+        ``preload=True`` or ``self.load_data()``.
```
Comment (drammock) [inline mne/evoked.py:169]: this is inaccurate, because Evoked objects are always loaded into memory.

File: mne/epochs.py [hunk 14]
```diff
@@ -1491,6 +1493,86 @@ def get_data(self, picks=None, item=None):
...
+    @verbose
+    def apply_function(self, fun, picks=None, dtype=None, n_jobs=1,
+                       channel_wise=True, verbose=None, *args, **kwargs):
+        """Apply a function to a subset of channels.
+
+        The function ``fun`` is applied to the channels defined in ``picks``.
+        The data of the Epochs object is modified in-place. If the function
+        returns a different data type (e.g. numpy.complex) it must be specified
+        using the dtype parameter, which causes the data type used for
+        representing the raw data to change.
```
Comment (drammock) [inline mne/epochs.py:1505]: there is a lot here that is redundant with the method you're adding to Evokeds, and redundant with the existing docstring for `Raw.apply_function()`.  Also the docstring parameters are now copy-pasted to 3 places in the codebase, which means changes potentially have to be made 3 times.  This is a perfect opportunity to learn to use our `docdict`!  Start by adding an entry to `mne/utils/docs.py` called `apply_function_summary` or so, that has this paragraph, and then other entries for the repeated function params.  Then come back here and insert them with the `%(docdict_key)s` method (like is already done for `%(picks_all_data_noref)` below).  For an example of how you can factor out the common elements of a paragraph, see the example of how we handle the various forms that the `picks` parameter can take: 

https://github.com/mne-tools/mne-python/blob/c638f74ac727aeaca3b9616e56115b990a84c7ba/mne/utils/docs.py#L439-L463

File: mne/epochs.py [hunk 15]
```diff
@@ -1491,6 +1493,86 @@ def get_data(self, picks=None, item=None):
...
+        %(verbose_meth)s
+        *args : list
+            Additional positional arguments to pass to fun (first pos. argument
+            of fun is the timeseries of a channel).
+        **kwargs : dict
+            Additional keyword arguments to pass to fun.
+
+        Returns
+        -------
+        self : instance of Raw
```
Comment (drammock) [inline mne/epochs.py:1544]:
```suggestion
        self : instance of Epochs
```

File: mne/tests/test_epochs.py [hunk 16]
```diff
@@ -3357,3 +3359,34 @@ def test_empty_constructor():
...
+                         l_trans_bandwidth=4, h_trans_bandwidth=4)
+    out_ch = epochs.apply_function(filter_data, channel_wise=True,
+                                   **filter_params)
+    out_ch_data = out_ch._data
+    # checking if the filtering was correctly applied
+    psd_out, freqs = psd_array_welch(out_ch_data[0, 0, :], sfreq=info['sfreq'])
+    freqs_up_out = freqs[psd_out > 1e-6]
+    psd_out_raw, freqs = psd_array_welch(data[0, 0, :], sfreq=info['sfreq'])
+    freqs_up_raw = freqs[psd_out_raw > 1e-6]
+    assert len(freqs_up_out) < len(freqs_up_raw)
```
Comment (drammock) [inline mne/tests/test_epochs.py:3392]: this seems unnecessarily complex way to test `apply_function()`.  Why not a super simple function that just flips the sign of every data value?  Easy to understand, easy to test.

File: mne/tests/test_evoked.py [hunk 17]
```diff
@@ -8,6 +8,7 @@
 from copy import deepcopy
 import os.path as op
 import pickle
+import mne
```
Comment (drammock) [inline mne/tests/test_evoked.py:11]: we don't usually import the whole `mne` module in our testing files, we just import the functions we need (see lines 19-27)

File: mne/tests/test_evoked.py [hunk 18]
```diff
@@ -759,3 +761,24 @@ def test_hilbert():
...
+
+
+def test_apply_function_evk():
+    """Check the apply_function method for evoked data."""
+    # create fake evoked data to use for checking apply_function
+    data = np.random.rand(2, 10, 1000)
+    events = np.array([[0, 0, 1], [INT32_MAX, 0, 2]])
+    info = mne.create_info(10, 1000., 'eeg')
+    epochs = mne.EpochsArray(data, info, events)
+    evoked = epochs.average(method='mean')
```
Comment (drammock) [inline mne/tests/test_evoked.py:780]: why not create the evoked directly using `EvokedArray`?  Also why do you need events here?

File: mne/tests/test_evoked.py [hunk 19]
```diff
@@ -759,3 +761,24 @@ def test_hilbert():
...
+    evoked_data = evoked.data
+    # check apply_function channel-wise
+    applied = evoked.apply_function(fun)
+    assert np.shape(applied.data) == np.shape(evoked_data)
```
Comment (drammock) [inline mne/tests/test_evoked.py:784]: can we test something more stringent besides just the data shape?  (like that the data values end up how we expect them to end up)?

File: mne/tests/test_epochs.py [hunk 20]
```diff
@@ -3357,3 +3359,34 @@ def test_empty_constructor():
...
+
+
+def fun(data_epochs):
+    """Auxiliary function to test_apply function."""
+    matrix = np.random.rand(10, 10)
+    return matrix @ data_epochs
+
+
+def test_apply_function():
+    """Test epochs takes into account repeated events."""
```
Comment (drammock) [inline mne/tests/test_epochs.py:3371]: the test description seems unrelated?

**Review by drammock — CHANGES_REQUESTED** [review]

File: examples/visualization/plot_sensor_noise_level.py [hunk 21]
```diff
@@ -29,3 +23,8 @@
 # We can plot the absolute noise levels:
 raw_erm.plot_psd(tmax=10., average=True, spatial_colors=False,
                  dB=False, xscale='log')
+###############################################################################
+# References
+# ----------
+#
+# .. footbibliography::
```
Comment (sappelhoff) [inline examples/visualization/plot_sensor_noise_level.py:30]: After doing a quick check of the other tutorials, there is usually no blank line here

```suggestion
# .. footbibliography::
```

File: mne/epochs.py (hunk 14 above)
Comment (vpeterson) [reply mne/epochs.py]: Great! I have done that now. Also for the `*args` and `**kwargs`. If agree on this, we should also change the raw function for consistency with the epochs and evoked objects?

**Comment by vpeterson** [comment]
Hi everyone! We think that we are ready for the final review with @nordme 
@drammock, let us know how the PR looks.

**Review by drammock — APPROVED** [review]
The tests look really nice now.

In lieu of extensive review comments on the docdict it turned out to be easier to push a commit.  I have done some docdict magic to remove even more redundancy. @vpeterson @nordme have a look at the changes on that commit and feel free to ask if you don't understand what I'm doing (or why).

**Comment by drammock** [comment]
@wmvanvliet can you look again and see if your request for changes is satisfied?
